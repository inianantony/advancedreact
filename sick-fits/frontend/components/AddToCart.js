import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { CURRENT_USER_QUERY } from './User'

const ADD_TO_CARD_MUTATION = gql`
    mutation ADD_TO_CARD_MUTATION($id:ID!){
        addToCart(id:$id){
            id
            quantity
        }
    }
`;

export default class AddToCart extends Component {
    render() {
        const { id } = this.props;
        return (
            <Mutation refetchQueries={[{ query: CURRENT_USER_QUERY }]} mutation={ADD_TO_CARD_MUTATION} variables={{ id }}>
                {(addToCart, { data, loading, error }) => {
                    return (
                        <button disabled={loading} onClick={addToCart}>Add{loading ? "ing" : ""} to Cart ðŸ›’</button>
                    );
                }}
            </Mutation>
        );
    }
}