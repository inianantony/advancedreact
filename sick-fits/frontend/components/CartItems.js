import React, { Component } from 'react';
import { Query, Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { ALL_ITEMS_QUERY } from './items';
import formatMoney from '../lib/formatMoney';
import styled from 'styled-components';
import { object } from 'prop-types';
import PropTypes from 'prop-types';

const CartItemStyled = styled.li`
    padding: 1rem 0;
    border-bottom: 1px solid ${props => props.theme.lightgrey};
    display: grid;
    align-items: center;
    grid-template-columns: auto 1fr auto;
    img{
        margin-right: 10px;
    }
    h3, p{
        margin: 0;
    }
`;

const CartItems = props => {
    const cartItem = props.cartItem;
    return (
        <CartItemStyled>
            <img width="100px" src={cartItem.item.image} alt={cartItem.item.title} />
            <div className={cartItem.item.image}>
                <h3>{cartItem.item.title}</h3>
                <p>
                    {formatMoney(cartItem.item.price * cartItem.quantity)}
                    {' - '}
                    <em>{cartItem.quantity} &times; {formatMoney(cartItem.item.price)} each</em>
                </p>
            </div>
        </CartItemStyled>
    );
}

CartItems.propTypes = {
    cartItem: PropTypes.object.isRequired
}

export default CartItems;