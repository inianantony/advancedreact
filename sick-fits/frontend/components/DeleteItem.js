import React, { Component } from 'react';
import { Query, Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { ALL_ITEMS_QUERY } from './items';

const DELETE_ITEM_MUTATION = gql`
    mutation DELETE_ITEM_MUTATION($id: ID!) {
        deleteItem(
            id: $id
        ){
            id
            title
            price
            description
        }
    }
`;

export default class DeleteItem extends Component {
    updateCache = (cache, payload) => {
        const data = cache.readQuery({ query: ALL_ITEMS_QUERY });
        console.log(data);
        data.items = data.items.filter(item => item.id !== payload.data.deleteItem.id);

        cache.writeQuery({ query: ALL_ITEMS_QUERY, data });
    }
    render() {
        return (
            <Mutation mutation={DELETE_ITEM_MUTATION} variables={{ id: this.props.id }} update={this.updateCache}>
                {(deletItem, { error }) => (
                    <button onClick={() => {
                        if (confirm('Are you sure to delete?'))
                            deletItem();
                    }}>{this.props.children}</button>
                )}
            </Mutation>
        );
    }

}