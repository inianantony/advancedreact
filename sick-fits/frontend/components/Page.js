import React from 'react'

export default class Page extends React.Component{
    render(){
        return (
            <div>
                <p> Hey ! I am the page Component</p>
                {this.props.children}
            </div>
        )
    }
}