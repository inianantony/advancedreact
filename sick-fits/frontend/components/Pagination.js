import React, { Component } from 'react';
import PaginationStyles from './styles/PaginationStyles';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';
import { perPage } from '../config'
import Head from 'next/head'
import Link from 'next/link'

const PAGINATION_QUERY = gql`
    query PAGINATION_QUERY {
        itemsConnection {
            aggregate{
                count
            }
        }
    }
`;

const Pagination = props => (

    <Query query={PAGINATION_QUERY}>
        {({ error, loading, data }) => {
            if (error) return <p>Error...</p>
            if (loading) return <p>Loading...</p>
            const count = data.itemsConnection.aggregate.count;
            const pages = Math.ceil(count / perPage);
            return (
                <PaginationStyles>
                    <Head>
                        <title>Sick Fits | Page {props.page} of {pages} </title>
                    </Head>
                    <Link prefetch href={{ path: './items', query: { page: props.page - 1 } }}>
                        <a aria-disabled={props.page<=1}>⬅ Prev</a>
                    </Link>
                    <p>Page {props.page} of {pages}</p>
                    <p>{count} Items Total</p>
                    <Link prefetch href={{ path: './items', query: { page: props.page + 1 } }}>
                        <a aria-disabled={props.page>=pages}>Next ➡</a>
                    </Link>
                </PaginationStyles>
            )
        }}
    </Query>


);

export default Pagination;