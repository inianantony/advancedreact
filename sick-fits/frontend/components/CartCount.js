import React, { Component } from 'react';
import { Query, Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { CURRENT_USER_QUERY } from './User';
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import CloseButton from './styles/CloseButton';
import SickButton from './styles/SickButton';
import User from './User';
import CartItems from './CartItems'
import calcTotalPrice from '../lib/calcTotalPrice';
import formatMoney from '../lib/formatMoney';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { TransitionGroup, CSSTransition } from 'react-transition-group'

const AnimationStyles = styled.span`
    position: relative;
    .count{
        display: block;
        position: relative;
        transition: all 1s;
        backface-visibility: hidden;
    }
    .count-enter{
        transform: scale(4) rotateX(0.5turn);
    }
    .count-enter-active{
        transform: rotateX(0);
    }
    .count-exit{
        top:0;
        position: absolute;
        transform: rotateX(0)
    }
    .count-exit-active{
        transform: scale(4) rotateX(0.5turn);
    }
`;

const Dot = styled.div`
    background: ${props => props.theme.red};
    color: white;
    border-radius: 50%;
    padding: 0.5rem;
    line-height: 2rem;
    min-width: 3rem;
    margin-left: 1rem;
    font-weight: 100;
    font-feature-settings: 'tnum';
    font-variant-numeric: tabular-nums;
`;

const CartCount = props => {
    const count = props.count;
    return (
        <AnimationStyles>
            <TransitionGroup>
                <CSSTransition unmountOnExit className="count" classNames="count" key={count} timeout={{ enter: 1000, exit: 1000 }}>
                    <Dot>{count}</Dot>
                </CSSTransition>
            </TransitionGroup>
        </AnimationStyles>
    )
}

export default CartCount;