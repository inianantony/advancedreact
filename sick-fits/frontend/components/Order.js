import React from 'react';
import CartStyles from './styles/CartStyles';
import Supreme from './styles/Supreme';
import CloseButton from './styles/CloseButton';
import SickButton from './styles/SickButton';
import { Query, Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import User from './User';
import CartItems from './CartItems'
import Error from './ErrorMessage'
import OrderStyles from './styles/OrderStyles'
import calcTotalPrice from '../lib/calcTotalPrice';
import formatMoney from '../lib/formatMoney';
import { adopt } from 'react-adopt';
import PropTypes from 'prop-types';
import { format } from 'date-fns';
import Head from 'next/head';

const SINGLE_ORDER_QUERY = gql`
    query SINGLE_ORDER_QUERY($id: ID!){
        order(id: $id){
            id
            charge
            total
            createdAt
            user{
                id
            }
            items{
                id,
                title
                description
                price
                image
                quantity
            }
        }
    }
`;

export default class Order extends React.Component {
    static propTypes = {
        id: PropTypes.string.isRequired
    }
    render() {
        return (
            <Query query={SINGLE_ORDER_QUERY} variables={{ id: this.props.id }}>
                {({ data, error, loading }) => {
                    if (error) return <Error error={error} />
                    if (loading) return <p>Loading...</p>
                    const order = data.order;
                    return <OrderStyles>
                        <Head>Sick Fits - Order {order.id}</Head>
                        <p><span> Order Id : </span><span>{this.props.id}</span></p>
                        <p><span> Charge : </span><span>{order.charge}</span></p>
                        <p><span> Date : </span><span>{format(order.createdAt, "MMMM d, YYYY h:mm a")}</span></p>
                        <p><span> Order Total : </span><span>{formatMoney(order.total)}</span></p>
                        <p><span> Item Count : </span><span>{order.items.length}</span></p>
                        <div className="items">
                            {order.items.map(item => (
                                <div className="order-item" key={item.id}>
                                    <img src={item.image} alt={item.title} />
                                    <div className="item-details">
                                        <h2>{item.title}</h2>
                                        <p>Qty: {item.quantity}</p>
                                        <p>Each: {formatMoney(item.price)}</p>
                                        <p>SubTotal: {formatMoney(item.price * item.quantity)}</p>
                                        <p>{item.description}</p>
                                    </div>
                                </div>
                            ))}
                        </div>
                    </OrderStyles>
                }}
            </Query>
        );
    }
}
