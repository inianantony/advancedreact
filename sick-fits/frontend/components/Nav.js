import Link from 'next/link';
import NavStyles from './styles/NavStyles'
import User from './User'
import SignOut from './SignOut'
import { Mutation } from 'react-apollo';
import { TOGGLE_CART_MUTATION } from '../components/Cart'
import CartCount from './CartCount';

const Nav = () => (
    <User>
        {
            ({ data: { me } }) => {
                return (
                    <NavStyles>
                        <Link href="/items">
                            <a>Shop</a>
                        </Link>
                        { me &&
                            (<>
                                <Link href="/sell">
                                    <a>Sell</a>
                                </Link>
                                <Link href="/orders">
                                    <a>Orders</a>
                                </Link>
                                <Link href="/me">
                                    <a>Account</a>
                                </Link>
                                <Mutation mutation={TOGGLE_CART_MUTATION}>{
                                    (toggleCart) => {
                                        return <button onClick={toggleCart}>
                                            My Cart
                                            <CartCount count={me.cart.reduce((total, item) => (total + item.quantity), 0)}></CartCount>
                                        </button>
                                    }}
                                </Mutation>
                                <SignOut />
                            </>)
                        }
                        {!me && (
                            <Link href="/signup">
                                <a>Sign In / Up</a>
                            </Link>
                        )
                        }
                    </NavStyles>
                )
            }
        }
    </User>
)

export default Nav;