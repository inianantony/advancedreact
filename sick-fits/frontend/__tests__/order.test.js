import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import Order, { SINGLE_ORDER_QUERY } from '../components/Order';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser, fakeCartItem, fakeOrder } from '../lib/testUtils';

const mocks = [
    {
        request: { query: CURRENT_USER_QUERY },
        result: { data: { me: { ...fakeUser(), cart: [fakeCartItem()], }, }, },
    },
    {
        request: { query: SINGLE_ORDER_QUERY, variables: { id: "ord123" } },
        result: { data: { order: { ...fakeOrder() }, }, },
    },
];

describe('<AddToCart/>', () => {
    it('renders and matches the snap shot', async () => {
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <Order id="ord123" />
            </MockedProvider>
        );
        await wait();
        wrapper.update();
        expect(toJSON(wrapper.find('OrderStyles'))).toMatchSnapshot();
    });
});