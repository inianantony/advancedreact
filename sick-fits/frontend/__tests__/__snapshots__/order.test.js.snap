// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddToCart/> renders and matches the snap shot 1`] = `
<OrderStyles>
  <div
    className="OrderStyles-sc-4oqalm-0 camMij"
  >
    <SideEffect(Head)>
      <Head />
    </SideEffect(Head)>
    <p>
      <span>
         Order Id : 
      </span>
      <span>
        ord123
      </span>
    </p>
    <p>
      <span>
         Charge : 
      </span>
      <span>
        ch_123
      </span>
    </p>
    <p>
      <span>
         Date : 
      </span>
      <span>
        April 1, 2018 8:00 AM
      </span>
    </p>
    <p>
      <span>
         Order Total : 
      </span>
      <span>
        $400
      </span>
    </p>
    <p>
      <span>
         Item Count : 
      </span>
      <span>
        2
      </span>
    </p>
    <div
      className="items"
    >
      <div
        className="order-item"
        key="0618965f-eea6-4ae8-b415-d3f23683393e"
      >
        <img
          alt="dolorum ab aliquam id sit delectus vel"
          src="illo.jpg"
        />
        <div
          className="item-details"
        >
          <h2>
            dolorum ab aliquam id sit delectus vel
          </h2>
          <p>
            Qty: 
            1
          </p>
          <p>
            Each: 
            $42.34
          </p>
          <p>
            SubTotal: 
            $42.34
          </p>
          <p>
            ipsa tempora aut sint aperiam occaecati illum
          </p>
        </div>
      </div>
      <div
        className="order-item"
        key="8d81098d-fec8-4184-b94b-e78c1398f8e7"
      >
        <img
          alt="temporibus aliquam sed sapiente soluta corporis commodi"
          src="similique.jpg"
        />
        <div
          className="item-details"
        >
          <h2>
            temporibus aliquam sed sapiente soluta corporis commodi
          </h2>
          <p>
            Qty: 
            1
          </p>
          <p>
            Each: 
            $42.34
          </p>
          <p>
            SubTotal: 
            $42.34
          </p>
          <p>
            eaque nisi cum ducimus nemo doloremque maiores
          </p>
        </div>
      </div>
    </div>
  </div>
</OrderStyles>
`;
